src/dials/algorithms/correlation/analysis.py

```diff
@@ -0,0 +1,334 @@
+from dials.algorithms.symmetry.cosym import CosymAnalysis
+from dials.algorithms.symmetry.cosym.plots import plot_coords, plot_rij_histogram
+from dials.util.exclude_images import get_selection_for_valid_image_ranges
+from dials.util.filter_reflections import filtered_arrays_from_experiments_reflections
+from dials.util.multi_dataset_handling import select_datasets_on_identifiers
+
+logger = logging.getLogger("dials.algorithms.correlation.analysis")
+
+phil_scope = iotbx.phil.parse(
+    """\
+partiality_threshold = 0.4
+  .type = float(value_min=0, value_max=1)
+  .help = "Use reflections with a partiality greater than the threshold."
+
+include scope dials.algorithms.symmetry.cosym.phil_scope
+
+relative_length_tolerance = 0.05
+  .type = float(value_min=0)
+  .help = "Datasets are only accepted if unit cell lengths fall within this relative tolerance of the median cell lengths."
+
+absolute_angle_tolerance = 2
+  .type = float(value_min=0)
+  .help = "Datasets are only accepted if unit cell angles fall within this absolute tolerance of the median cell angles."
+
+min_reflections = 10
+  .type = int(value_min=1)
+  .help = "The minimum number of reflections per experiment."
+""",
+    process_includes=True,
+)
+
+
+class CorrelationMatrix:
+    def __init__(
+        self,
+        experiments: ExperimentList,
+        reflections: list,
+        params: scope_extract = None,
+    ):
+        """
+        Set up the required cosym preparations for determining the correlation matricies
+        of a series of input experiments and reflections.
+
+        Args:
+            experiments (dxtbx_model_ext.ExperimentList): list of experiments in dials format
+            reflections (list): list of dials_array_family_flex_ext.reflection_table objects associated with the experiments
+            params (libtbx.phil.scope_extract):
+        """
+
+        # Set up experiments, reflections and params
+
+        if params is None:
+            params = phil_scope.extract()
+        self.params = params
+        self._reflections = []
+
+        if len(reflections) == len(experiments):
+            for refl, expt in zip(reflections, experiments):
+                sel = get_selection_for_valid_image_ranges(refl, expt)
+                self._reflections.append(refl.select(sel))
+        else:
+            sys.exit("Number of reflections and experiments do not match.")
+
+        # Initial filtering to remove experiments and reflections that do not meet the minimum number of reflections required (params.min_reflections)
+
+        self._experiments, self._reflections = self._filter_min_reflections(
+            experiments, self._reflections
+        )
+
+        # Used for optional json creation that is in a format friendly for import and analysis (for future development)
+        self.ids_to_identifiers_map = {}
+        for table in self._reflections:
+            self.ids_to_identifiers_map.update(table.experiment_identifiers())
```

Comment:
I'm not quite sure, because I've not tested this in anger with multiple reflection tables, but it looks at a glance like each `table.experiment_identifiers()` is going to look like {0: ..., 1: ..., ...}, so each `.update` is going to discover the same keys again and again and we're going to end up with a dictionary `self.ids_to_identifiers_map` with only as many keys as the longest `table.experiment_identifiers()`, and with several of the values overwritten by the last occurring value for a given key, rather than a dictionary containing all the experiment identifiers from all the reflection tables.  But it may be that I don't have a good understanding of how `reflection_table.experiment_identifiers` works and so have misinterpreted...

Reply:
After a bit of testing with a dataset, it looks like each table only has one identifier (and associated id), so `table.experiment_identifiers()` seems to return only a single identifier, so nothing seems to get overwritten. Ie the first `table` has `table.experiment_identifers()` returning `{0: .....}`, and the second returning `{1: .....}`, and so forth. So unsure why `table.experiment_identifiers()` is plural, since it only seems to be returning one `key:identifier` pair for each, unless that's just an artefact of how we've set up the reflection tables.

Reply:
OK, thanks for checking.  In that case, I'm happy on this score.